[ECE] Add ECE support for WooCommerce Deposits.

asumaran opened the pull request:
Closes https://github.com/Automattic/woocommerce-payments/issues/9067

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

_Based on testing instructions of https://github.com/Automattic/woocommerce-payments/pull/7910 which added support for WooCommerce Deposits with PRBs. The functionality should be identical with ECE._


- Enable Express Checkouts (Google and Apple Pay) to be displayed on product pages.
- Install and enable WooCommerce Deposits. You can download it from [here](https://woocommerce.com/my-account/downloads/)

**Virtual Product**
- Create a virtual simple product with an optional deposit of 50%
- <img width="375.5" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/1025173/34a7be76-f574-43fe-adec-446151b69475">
- Visit virtual simple product with optional deposit of 50%
- Select "Pay Deposit" (if not already selected)
- Click Apple/Google Pay button.
- The price in the payment sheet should reflect the deposit amount (Only the 50% of the total amount in this case).
- Proceed with purchase.
- The deposit amount should be recorded as paid against the order.
- The remaining amount should be recorded against the order.

**Physical Product**
- Create a simple product with optional deposit of 30%
- <img width="378" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/1025173/f476b00c-298d-4d43-8d91-9e72a675f632">
- Visit the product page then click the express checkout button.
- Check the paid amount is only the 30% of the total `(product price + shipping) * 0.3`


**Virtual product with payment plan**
- Go to Products > Payment Plans and create a 30%/70% payment plan.
- <img width="303" alt="Screenshot 2024-07-10 at 8 53 18 PM" src="https://github.com/Automattic/woocommerce-payments/assets/1025173/445e44f0-245b-478e-9fcc-d9b922ffcaab">
- Create a virtual product with optional plan
- <img width="827" alt="Screenshot 2024-07-10 at 8 52 17 PM" src="https://github.com/Automattic/woocommerce-payments/assets/1025173/755b1e6c-2333-4280-95b3-81bd8c18524f">
- Go to the product page.
- Select "Pay Deposit".
- Select payment plan.
- Click Apple/Google Pay button.
- Proceed with purchase.
- The first payment amount should be recorded as paid against the order.
- The remaining amount should be recorded against the order.

**Note**
If you want to test to pay the remaining balance you have to go to the order page in the admin area then click "Invoice Remaining Balance"

<img width="525" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/1025173/a56df1f9-838d-40ae-a86e-1b235e7a5767">

It will create a new order, and you will be able to pay the remaining using the "Customer payment page" link. Note that the ECE button didn't work on the Pay for Order page, so code has been added to disable it on that page in this PR.

<img width="516" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/1025173/a722dd8f-ab56-4214-893e-d14078e24620">


-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)


**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

asumaran committed:
Support deposits with variations

asumaran committed:
Disable ECE for follow up orders on the Pay For Order page

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9081` or branch name `as-ece-wc-deposits-support` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=as-ece-wc-deposits-support) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: c519b21018726215083f5583a0a8c7ec0bd39fb1
- Build time: 2024-07-16 13:29:22 UTC

_Note: the build is updated when a new commit is pushed to this PR._

asumaran committed:
Add changelog entry

reykjalin commented on client/express-checkout/index.js:
```diff
@@ -487,6 +487,20 @@ jQuery( ( $ ) => {
+			$(
+				'input[name=wc_deposit_option],input[name=wc_deposit_payment_plan]'
+			)
+				.off( 'change' )
```
What's the purpose of this `off`? Do we need to remove all other event handlers from the change event? 🤔

asumaran replied:
I was being extra cautious because the ECE element can be reinitialized, event handlers are attached each time it happens. To avoid attaching multiple handlers to the same selector, I remove any existing event handlers first. Although the elements are not being removed and added again, it seems unnecessary, but I prefer to keep it this way for safety. Does that sound good to you?

reykjalin replied:
I appreciate the caution, but if we're going to do that I think we should pass the specific handler we're removing to the function so we're not just removing every single event handler that may have been attached.

See the [off()](https://api.jquery.com/off/) docs for details.

reykjalin commented on includes/express-checkout/class-wc-payments-express-checkout-button-helper.php:
```diff
@@ -394,6 +394,13 @@ public function should_show_express_checkout_button() {
+			$order = wc_get_order( absint( get_query_var( 'order-pay' ) ) );
+			if ( $order && class_exists( 'WC_Deposits_Order_Manager' ) && WC_Deposits_Order_Manager::is_follow_up_order( $order ) ) {
+				return false;
+			}
```
Comment on the comment here: Is this specifically that ECE doesn't work _for WC Deposits_ on the Pay for Order (PfO) page, or in general? I saw the conversation in Slack where you figured out a general solution to ECE on the PfO page? 🤔

asumaran replied:
>Is this specifically that ECE doesn't work for WC Deposits on the Pay for Order (PfO) page, or in general? 

I thought it only failed for WC Deposits but it failed in general **for me**. My bad. I've confirmed the issue happened because the Cart Testing protection service was enabled. It'll be fixed on https://github.com/Automattic/woocommerce-payments/pull/9089

asumaran commented on includes/express-checkout/class-wc-payments-express-checkout-button-helper.php:
```diff
@@ -394,6 +394,13 @@ public function should_show_express_checkout_button() {
 		if ( $this->is_pay_for_order_page() ) {
+
+			// ECE doesn't work on the Pay For Order Page.
+			$order = wc_get_order( absint( get_query_var( 'order-pay' ) ) );
```
Actually, looks like there was an issue with my local install and might have fixed it on https://github.com/Automattic/woocommerce-payments/issues/9089

I had Card Testing Prevention enabled on my local site and it seems it was causing ECE to fail on the Pay for Order page. 

I'll double check tomorrow.

reykjalin reviewed (changes requested):
The payments failed for me in several places which I think we need to address:

## Virtual Simple (50% deposit)

- ⛔️ Fails when using GPay from the product page.
- ✅ Success when using GPay from the Cart Block.
- ⛔️ Fails when using GPay from the Checkout Block.
- ✅ Success when using GPay from the Shortcode Checkout.
- ✅ Success when using GPay from the Shortcode Cart

## Physical product (30% then 70%)

- ⛔️ Payment sheet doesn't show up from the product page (default choice is **Pay full**, change to **Deposit** and click button).
- ✅ Success when using GPay from the Cart Block.
- ⛔️ Fails when using GPay from the Checkout Block.
- ✅ Success when using GPay from the Shortcode Checkout.
- ✅ Success when using GPay from the Shortcode Cart

## Payment Plan (physical)

- ✅ Success when using GPay from the product page.
    - I have no idea why this one succeeded but the other ones failed.
- ✅ Success when using GPay from the Cart Block.
- ⛔️ Fails when using GPay from the Checkout Block.
- ✅ Success when using GPay from the Shortcode Checkout.
- ✅ Success when using GPay from the Shortcode Cart

asumaran committed:
Remove unnecessary code

asumaran committed:
Merge branch 'develop' into as-ece-wc-deposits-support

asumaran committed:
Move declaration to the top

asumaran committed:
Merge branch 'develop' into as-ece-wc-deposits-support

asumaran committed:
Remove incorrect paramenter

asumaran committed:
Reinitialize ECE for deposits

We need to do this because the accumulated amount in the item list is higher than the total passed.

asumaran committed:
Merge branch 'develop' into as-ece-wc-deposits-support

asumaran commented:
>⛔️ Fails when using GPay from the product page.

Nice catch. Fixed via [2203c28](https://github.com/Automattic/woocommerce-payments/pull/9081/commits/2203c2842a35a2d9dd77187005f859e2f6327f8e)

> ⛔️ Fails when using GPay from the Checkout Block.

Works for me now.

> ⛔️ Payment sheet doesn't show up from the product page (default choice is Pay full, change to Deposit and click button).

I can reproduce this. I've fixed it via https://github.com/Automattic/woocommerce-payments/pull/9081/commits/263027ce028d9d7b228fad8ab785f5b02b5ed941. Race condition issue.

> ⛔️ Fails when using GPay from the Checkout Block.

Works for me now.

> ⛔️ Fails when using GPay from the Checkout Block.

Works for me now.

reykjalin reviewed (changes requested):
I'm still seeing some failures in the Checkout block:

## Virtual Simple (50% deposit)

- ✅ Success when using GPay from the product page.
- ✅ Success when using GPay from the Cart Block.
- ⛔️ Fails when using GPay from the Checkout Block.
- ✅ Success when using GPay from the Shortcode Checkout.
- ✅ Success when using GPay from the Shortcode Cart

## Physical product (30% then 70%)

- ✅ Success sheet doesn't show up from the product page (default choice is **Pay full**, change to **Deposit** and click button).
- ✅ Success when using GPay from the Cart Block.
- ⛔️ Fails when using GPay from the Checkout Block.
- ✅ Success when using GPay from the Shortcode Checkout.
- ✅ Success when using GPay from the Shortcode Cart

## Payment Plan (physical)

- ✅ Success when using GPay from the product page.
    - I have no idea why this one succeeded but the other ones failed.
- ✅ Success when using GPay from the Cart Block.
- ⛔️ Fails when using GPay from the Checkout Block.
- ✅ Success when using GPay from the Shortcode Checkout.
- ✅ Success when using GPay from the Shortcode Cart

All I can tell from the network console is that it's failing in the `create_order` part of the flow. Maybe because I have multi-currency enabled? 🤔  Although I doubt that's the problem since this works via the same APIs from other pages. I suspect there's something that needs to be changed in the blocks integration?

asumaran commented:
@reykjalin does it fail for you with PRBs too?

rafaelzaleski committed:
Merge branch 'develop' into as-ece-wc-deposits-support

rafaelzaleski commented:
I couldn’t reproduce the [failed test cases](https://github.com/Automattic/woocommerce-payments/pull/9081#pullrequestreview-2176696246) reported by @reykjalin in Safari. 
I’ve deployed this branch to the Pressable test environments we’re using today, as it includes the WC Deposits integration. 

Could @reykjalin check if the issue still persists in the Pressable environment? This could help us understand why it’s failing for Kris and determine whether to address it here or in a new issue, especially if it’s specific to certain setups.

asumaran commented:
I reran some tests with the failed cases, and it's working fine for me on Chrome. I will need more details to reproduce it, or maybe it's related to your setup, @reykjalin?

reykjalin commented:
> Could @reykjalin check if the issue still persists in the Pressable environment?

They're not present there so I'll do some more testing in my local tomorrow to try to track this down.

asumaran committed:
Do not ask for shipping if no shipping options are created

asumaran committed:
Merge branch 'develop' into as-ece-wc-deposits-support

reykjalin reviewed (approved):
The problems I was running into are unrelated to the changes in this PR so I'm approving the PR given that it seems to work for everyone else, including for myself in a different testing environment.

Awesome work 👏 🚀
